keras/applications/convnext.py

```diff
@@ -0,0 +1,660 @@
+      specified.
+    classifier_activation: A `str` or callable. The activation function to use
+      on the "top" layer. Ignored unless `include_top=True`. Set
+      `classifier_activation=None` to return the logits of the "top" layer.
+
+  Returns:
+    A `keras.Model` instance.
+
+  Raises:
+      ValueError: in case of invalid argument for `weights`,
+        or invalid input shape.
+      ValueError: if `classifier_activation` is not `softmax`, or `None` 
+        when using a pretrained top layer.
+      ValueError: if `include_top` is True but `num_classes` is not 1000 
+        when using ImageNet.
+  """
+  if not (weights in {"imagenet", None} or tf.io.gfile.exists(weights)):
+    raise ValueError("The `weights` argument should be either "
+                     "`None` (random initialization), `imagenet` "
+                     "(pre-training on ImageNet), "
+                     "or the path to the weights file to be loaded.")
+
+  if weights == "imagenet" and include_top and classes != 1000:
+    raise ValueError("If using `weights` as `'imagenet'` with `include_top`"
+                     " as true, `classes` should be 1000")
+
+  # Determine proper input shape.
+  input_shape = imagenet_utils.obtain_input_shape(
+    input_shape,
+    default_size=default_size,
+    min_size=32,
+    data_format=backend.image_data_format(),
+    require_flatten=include_top,
+    weights=weights)
+
+  if input_tensor is None:
+    img_input = layers.Input(shape=input_shape)
+  else:
+    if not backend.is_keras_tensor(input_tensor):
+      img_input = layers.Input(tensor=input_tensor, shape=input_shape)
+    else:
+      img_input = input_tensor
+
+  if input_tensor is not None:
+    inputs = utils.layer_utils.get_source_inputs(input_tensor)
+  else:
+    inputs = img_input
+
+  x = inputs
+  if include_preprocessing:
+    x = PreStem(name=model_name)(x)
+  
+  # Stem block.
+  stem = tf.keras.Sequential(
+    [
+      layers.Conv2D(projection_dims[0], kernel_size=4, strides=4,
+        name=model_name + "_stem_conv"),
+      layers.LayerNormalization(epsilon=1e-6, 
+        name=model_name + "_stem_layernorm"),
+    ],
+    name=model_name + "_stem",
+  )
+
+  # Downsampling blocks.
+  downsample_layers = []
+  downsample_layers.append(stem)
+  for i in range(3):
+    downsample_layer = tf.keras.Sequential(
+      [
+        layers.LayerNormalization(epsilon=1e-6, 
+          name=model_name + "_downsampling_layernorm_" + str(i)),
+        layers.Conv2D(projection_dims[i + 1], kernel_size=2, strides=2,
+          name=model_name + "_downsampling_conv_" + str(i)),
+      ],
+      name=model_name + "_downsampling_block_" + str(i),
+    )
+    downsample_layers.append(downsample_layer)
+  
+  # Stochastic depth schedule.
+  dp_rates = [x for x in tf.linspace(0.0, drop_path_rate, sum(depths))]
+
+  # ConvNeXt stages.
+  stages = []
+  cur = 0
+  for i in range(4):
+    stage = tf.keras.Sequential(
```

Comment:
Prefer using the Functional API here to do instantiation + application at the same time.

Reply:
Sure. I thought it was more readable. If it was done using Functional API it would have looked something like so:

```py


for i in range(len(4)):
    x = downsample_layers[i](x)
    for j in range(depths[i]):
        x = ConvNeXtBlock(
            projection_dim=projection_dims[i],
            drop_path_rate=dp_rates[cur + j],
            layer_scale_init_value=layer_scale_init_value,
            name=model_name + f"_stage_{i}_block_{j}",
        )(x)
    cur += depths[i]
```

Reply:
@fchollet a gentle reminder.